### ISSUE TITLE
Automatic library loading (`load-library`)

### ISSUE THREAD

**Issue by apiraino** [yjwen/org-reveal#445]
Hi,

I am trying to automate the library loading by putting it into my init.el
I'm trying this code to initialize the export backend:
```
(load-library "ox-reveal")
(use-package ox-reveal
  :init
  (add-hook 'after-save-hook (lambda () (when (eq 'org-mode major-mode)
                                          (org-reveal-export-to-html-and-browse))))
  )
```

How when I save the buffer, the export command fails with:
`org-export-barf-if-invalid-backend: Unknown "nil" back-end: Aborting export`

If I manually run `M-x load-library RET ox-reveal` then the backend is loaded and the export works.

Same behaviour if I use `(require 'ox-reveal)`.

I'm wondering if I'm doing something wrong :-)

Thanks for any suggestion!

**Comment by yjwen** [comment]
I tried with your code in my .emacs, both the `(load-library "ox-reveal")` and `(require 'ox-reveal)` versions, but didn't see that error. What is the version of use-package that produced that error? I am using use-package 20210202.1926 from MELPA.

**Comment by apiraino** [comment]
here's the relevant packages:
- use-package 20210207.1926 from MELPA
- ox-reveal 20210515.1007 from MELPA

I did a bit of debug (it it helps).

- launched emacs with `--debug-init` and I see no error
- `M-x toggle-debug-on-error` after emptying my .elc cache
- apply a change to my slides, then `C-x C-s` and I see this:

```
Debugger entered--Lisp error: (error "Unknown \"nil\" back-end: Aborting export")
  signal(error ("Unknown \"nil\" back-end: Aborting export"))
  error("Unknown \"%s\" back-end: Aborting export" nil)
  org-export-barf-if-invalid-backend(nil)
  org-export-as(reveal nil nil nil (:output-file "slides.html"))
  org-export-to-file(reveal "slides.html" nil nil nil nil nil)
  org-reveal-export-to-html(nil nil nil nil nil)
  org-reveal-export-to-html-and-browse()
  (progn (org-reveal-export-to-html-and-browse))
  (if (eq 'org-mode major-mode) (progn (org-reveal-export-to-html-and-browse)))
  (lambda nil (if (eq 'org-mode major-mode) (progn (org-reveal-export-to-html-and-browse))))()
  run-hooks(after-save-hook)
  basic-save-buffer(t)
  save-buffer(1)
  funcall-interactively(save-buffer 1)
  call-interactively(save-buffer nil nil)
  command-execute(save-buffer)
```

Another thing (not sure if relevant): the reveal-js export backend is available only after loading the library (which is ok) but even after loading the library, if I check the available backends with `M-x customize-varilable RET org-export-backends` I dont't see it.

**Comment by yjwen** [comment]
`org-export-backends` is the list of org-export's default backends. It is OK `reveal` is not there, as it is not part of the orgmode distribution. But once `ox-reveal` is successfully loaded, the variable `org-export-registered-backends` should have an entry of `:name reveal`. Could you please check the content of `org-export-registered-backends`?

**Comment by apiraino** [comment]
yay! @yjwen I've found the reason, it was my configuration.

I had configured Org mode to only have a subset of export backends (I don't need all of them), so in my `custom.el` I have:
```
'(org-export-backends '(html md odt))
```
and this is preventing org-reveal to install its backend. If I comment that line, org-reveal loads and works just fine.

As you correctly point out I can't add org-reveal to the list of default backends, but I also don't see anywhere the variable `org-export-registered-backends` (also tried a fresh install of emacs to be sure).

I don't how how to read the above information, if it's a bug in org-reveal or was simply a problem in my config. I've updated the previous message mentioning the org mode version I have (9.4.5) in case it has a role.

**Comment by yjwen** [comment]
I see. I finally reproduced your issue by your information. The problem is not which backend is set in `org-export-backends`, but the customization to `org-export-backends` itself. I customized that variable to its default value `'(ascii html icalendar latex odt)`, the issue still exists. Only when the customization is cleared will the issue disappear. I'd prefer to attribute the issue to org-mode.

`org-export-registered-backends` is not a customizable variable, so you cannot `M-x customize-variable` to it. But you can check its current value by `C-h C-v org-export-registered-backends`.

**Comment by apiraino** [comment]
ah thank you so much for explaining, now I have a better picture :) I'll try submitting this topic to the emacs mailing llist and get a feedback.

I'll close this issue (if it's ok for you) and hopefully report back at some point

**Comment by yjwen** [comment]
No problem
